mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp

```diff
@@ -949,6 +949,145 @@ mlir::scf::tileAndFuseProducerOfSlice(
                                            tileAndFuseResult->tiledOps};
 }
 
+/// Get the real producer from candidate ExtractSliceOp
+///
+/// ```
+/// %0 = producer
+/// %1 = scf.for(%arg1 = %0)
+///   %2 = extract %arg1
+///   %3 = scf.for(%arg2 = %2)
+///      %4 = extract %args2
+///      ...
+/// ```
+///
+/// @param candidateSliceOp: %4 = extract %args2
+/// @param backwardSlice: in-out parameter populated by backward extractSliceOps
+/// @return OpResult Producer : %0 = producer
+static FailureOr<OpResult> getRealProducerFromExtractSliceOp(
```

Comment:
Instead of doing this, would it help if you just folded the `extract_slices` using something like [this](https://github.com/iree-org/llvm-project/blob/c9f6518f742c88bda309d5331e0a5d4664387f94/mlir/lib/Dialect/Tensor/Transforms/MergeConsecutiveInsertExtractSlicePatterns.cpp#L23) pattern

Reply:
I know that we can merge consecutive `extractSlice`. But, I am sorry that we have to just keep these consecutive `extractSlice` for selection. The most intuitive example is as below:

```
%0 = producer
%1 = scf.for(%arg1 = %0)
   %2 = extract %arg1
   %3 = scf.for(%arg2 = %2)
      %4 = extract %args2
```
In some cases, the producer could be only fused into the outer candidate slice due to semantic of producer, i.e. reduce/pack/unpack those ops have specific demands of tiling size. If we merge the two consecutive `extractSlice`, it may break this fusion. And that is why we emphasize multi-level candidates here.

Reply:
That is OK, but the point here is to fuse with `%2` you dont need to look through `extract_slices`.... Just not sure of why there would be a case where you need to look through extract_slices to get to the producer, when you cant collapse all the extract slices to begin with.

Reply:
That is just simple test. If we have one more candidate slice, saying:

```
%0 = producer
%1 = scf.for(%arg1 = %0)
   %2 = extract %arg1
   %3 = scf.for(%arg2 = %2)
      %4 = extract %args2
      %5 = scf.for(%arg3 = %4)
          %6 = extract %args3
          tiled_consumer ins(%6)
```
where `%2` and `%4` are both valid candidate to fuse producer. And the user just want to manually pass `%4` to `tileAndFuseProducerOfSlice` because `%4` usually has smaller tileSize in most cases. Then, the current `getUntiledProducerFromSliceSource` will fail to get real producer without looking through `extract_slices`.

Another use case is that if someone want to enable automatic fusion by starting with `tiled_consumer` just like what `tileConsumerAndFuseProducersUsingSCF` did. The available extract slice we can find via operand of `tiled_consumer` is only the nearest candidate `%6`. Then, we also need to look through all `extract_slices` to get real producer.

Meanwhile, we dont know which candidate slice could be fuse until we know what the exact producer is. Thus, it is unreasonable to collapse all the extract slices before fusion.

Reply:
I am not sure this reasoning holds. If you want to fuse `%4` and not `%2` you are effectively "fusing" the slices and the doing the producer fusion. It would be better to stage it so that you fuse the slices so that you get the producer directly and then do the fusion. So effectively this change is trying to do both of this
1. Combine extract slices
2. Do producer fusion
in code. That is just added complexity. It would be easier to just combine the extract slices first and then fuse the producer.

Reply:
Let me try to rephrase the whole process step by step:
1. look through to collect all backward candidate slice of tiled operand `%6`, including `%6 = extract %args3`, ` %4 = extract %args2`, `%2 = extract %arg1`.
2. select and filter which candidate to fuse. I.e. ` %4 = extract %args2`.
3. Iteratively do producer fusion, from `%2` to `%4` by two iterations.

Through second thought, I agree that we can replace `step3` with your advice, which firstly combined `%2` and `%4`, then execute only single iteration of producer fusion.

However, as for `step 1` and `step 2`, we need to look through extract slice anyway... As I explained early, we could not fold `%2` and `%4` before these two steps. 

BTW, I dont think original `step3` in iterative fashion adding too much complexity because it is just reusing same logic and code.

On the other side, merging two extract slice in our case seems a little different from [MergeConsecutiveExtractSlice](https://github.com/iree-org/llvm-project/blob/c9f6518f742c88bda309d5331e0a5d4664387f94/mlir/lib/Dialect/Tensor/Transforms/MergeConsecutiveInsertExtractSlicePatterns.cpp#L23) as you pointed out, which actually expects direct use-def relationship:

```
%0 = extract_slice
%1 = extract_slice %0
```
rather than indirect relationship in our example:

```
%0 = extract_slice
scf.for(%1=%0)
  %2 = extract_slice %1
```

In this case, it is also necessary to look through extract slice before we merge these two slices.

Reply:
```
%0 = producerOp
scf.for(%1=%0)
   %2 = extract_slice %1
   scf.for(%3=%2)
      %4 = extract_slice %3
```

Besides, I am worried about merging two candidates interleaved by a `scf.for` with its `inits` involved is not trivial...

Reply:
Hi, @MaheshRavishankar. Since PR #108318 has been broken down into three parts:
1. support single nested `scf.for` (merged, great thanks for your review!)
2. multi-level candidates
3. multiple users

I think the second part is similar to this patch, no matter producer or consumer fusion. In general, this patch focus on how to deal with consecutive candidates **interleaved by a `scf.for`**.

Lets look back what the major argument remains so far to speed up our next stage of review:
1. why need to look through the total chain of candidate slice?
2. why not merge two consecutive candidates?

The answer is listed in above several threads. The root cause behind are two different solutions to solve multi-level candidates cases:
1. merge consecutive candidates in advance.
2. iteratively fuse producer/consumer with one of candidates step by step(from outer to inner).

IMO, the most concern about the first solution is that it may introduce too many additional transform regarding `scf.for`. E.g..

```
%0 = producerOp
scf.for(%1=%0)
   %2 = extract_slice %1
   scf.for(%3=%2)
      %4 = extract_slice %3
      ... 
      yield %x
```

As I explained above, this scenario is a little different from what [`MergeConsecutiveExtractSlice`](https://github.com/iree-org/llvm-project/blob/c9f6518f742c88bda309d5331e0a5d4664387f94/mlir/lib/Dialect/Tensor/Transforms/MergeConsecutiveInsertExtractSlicePatterns.cpp#L23) expects due to `scf.for`. 

If we just merge consecutive extract slice(BTW, here we also need look through to penetrate `scf.for` ), the possible resultant IR looks like below:
```
%0 = producerOp
scf.for(%1=%0)
   %2 = extract_slice %1
   scf.for(%3=%2)
      %4 = extract_slice %1 [newOffsets] [newSizes]
      ... 
      yield %x
```
Then the problem is that how we modify new `inits` of `scf.for` with latest `dpsInit` of fused `producerOp`? ([`getUntiledProducerFromSliceSource`](https://github.com/llvm/llvm-project/blob/main/mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp#L1058) will fail to find `destinationIterArg`)

On the other hand, if we modify `%3=%2` at the same time, the resultant IR would become:
```
%0 = producerOp
scf.for(%1=%0)
   %2 = extract_slice %1
   scf.for(%3=%1)
      %4 = extract_slice %3 [newOffsets] [newSizes]
      ... 
      yield %x_new
```
First of all, this transform has already introduced more code change and there is an assumption that `%3` has only single one user. Moreover, considering the semantic of `scf.for`, it seems that we also need to modify yield value `%x`(and even its producer, like `insert_slice`) to make the result of `scf.for` matched with its `inits` . 

In summary, I am afraid that this kind of complexity is much more than the second solution based on simple iterative process of reusing existed function.

Reply:
Thanka for the explanation. I see the issues a bit more clearly now. I havent reviewed the code yet. I dont know if this has been rebased on top of what was submitted. 
But here is the complexity that I am concerned about. In you example, you need to somewhere keep track of the sequence of extract slices that you need to walk to get the producer because the actual offset and size you need is obtained by "combining" the offsets and sizes of all the slices to get the "real offset" and size.  Also I am not convinced that fusing the first extract slice + producer and then doing the second extract slice + producer is not feasible. That should be always possible.

So if you start with this

```
%0 = producerOp
scf.for(%1=%0)
   %2 = extract_slice %1
   scf.for(%3=%2)
      %4 = extract_slice %3
      ... 
      yield %x
```

you should always be able to do

```
scf.for(%1=%0)
   %2 = tiled producer 1
   scf.for(%3=%2)
      %4 = extract_slice %3
      ... 
      yield %x
```

and then you do 

```
scf.for(%1=%0)
   scf.for(%3=%2)
      %4 = tiled producer 2
      ... 
      yield %x
```

The amount of state you need to carry during the transformation to fuse with one extract slice is prohibitively high IMO. That will make it hard to change/fix the transformation.

Reply:
> I dont know if this has been rebased on top of what was submitted.

Not rebased yet, because I think it is more import to reach agreement on solution before pushing...

> you need to somewhere keep track of the sequence of extract slices that you need to walk to get the producer because the actual offset and size you need is obtained by "combining" the offsets and sizes of all the slices to get the "real offset" and size.

Yes, from the view of consequentialism, the real offset is indeed combined. However, there are something else we need to address, like what I explained above regarding loop transform. In this way, each two of extract_slice  (interleaved by a `scf.for`) merged, the certain `scf.for`(and its yieldValue, etc...) need transformed. Is the amount of state need to carry similar to what I do in iterative fashion?

> Also I am not convinced that fusing the first extract slice + producer and then doing the second extract slice + producer is not feasible. That should be always possible.

It is decided by concrete semantic of tilable op, such as `reduce`/`pack`/`unpack` op. Lets say fusing producer `unpack` into following loop:

```
// unpack tensor from ABab to AB
%1 = tensor.unpack ... inner_tiles = [32, 32] ... : tensor<2x2x32x32xbf16> -> tensor<64x64xbf16>
scf.for(0, 2) { // tile A dimension
 extract_slice1 // tileSize<32, 64>
 scf.for(0, 4) { // tile B dimension
    extract_slice2 // tileSize<32, 16>
    ...
 }
}
```
As you can see, the `tileSize` comes from `extract_slice2` is `<32,16>`, but `tensor.unpack` prefer perfect tiling case, i.e. tileSize should be exactly divided by `inner_tiles`. So, it may be not feasible in this case.

BTW, fusing consumer `reduce` maybe more intuitive: 

```
%0 = scf.for(0, 2) { // tile 128 dimension
 scf.for(0, 4) { // tile B dimension
    ...
   insert_slice1 // tileSize<64, 64>
 }
 insert_slice1 // tileSize<64, 256>
}
%2 = linalg.reduce { arith.addf } ins(%0 : tensor<128x256xf32>) outs(%1: tensor<128xf32>) dimensions = [1]
```
We could not furtherly fuse `reduce` into the inner-most insert slice, otherwise, it will lead to partial reduce(its another topic).


> So if you start with this

> you should always be able to do

> and then you do

Yes, it is exactly what I do now. IIRC, that is also what you suggest before in nested consumer fusion...

Reply:
Ping